cast/src/lib.rs

```diff
@@ -623,6 +654,31 @@ where
     }
 }
 
+pub struct RawLogStream<'a, M, R, T> {
+    provider: M,
+    stream: Option<FilterWatcher<'a, T, R>>,
+    logs_filter: Filter,
+}
+
+impl<'a, M: Middleware, R, T> RawLogStream<'a, M, R, T>
+where
+    M::Error: 'static,
+{
+    pub fn new(logs_filter: Filter, provider: M, stream: Option<FilterWatcher<'a, T, R>>) -> Self {
+        Self { provider, logs_filter, stream }
+    }
+
+    pub async fn next(&self) -> Result<Option<Vec<Log>>> {
+        if let Some(stream) = self.stream {
+            if let Some(block) = stream.next().await {
+                self.logs_filter = self.logs_filter.at_block_hash(block);
+            }
+        }
+        let logs = self.provider.get_logs(&self.logs_filter).await?;
+        Ok(Some(logs))
+    }
```

Comment:
lets implement https://docs.rs/futures/latest/futures/stream/trait.Stream.html instead and make this type an actual `Stream`

instead of `await`ing you need to `.poll` the streams

See also https://tokio.rs/tokio/tutorial/streams